### PR TITLE
Plugins: Filter Plugins During Search and Install

### PR DESCRIPTION
**Pull request by clintoncwolfe**
This PR:
 * Adds a JSON config file at etc/plugin_filters.json, which lists the names of Gems that we wish to exclude from plugin activity
 * Each exclusion must have a Rationale, which is used in error messages
 * `inspec plugin search` uses the exclusion list to filter results
 * `inspec plugin install` refuses to install gems on the exclusion list, and explains why in detail
 * `inspec plugin search` now hides `{inspec,train}-test-fixture` by default (using a different mechanism), but will reveal it when passed a special flag.

### CODE REVIEW

File: lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb [hunk 1]
```diff
@@ -35,16 +36,31 @@ def list
...
+      desc 'search [options] PATTERN', 'Searches rubygems.org for plugins.'
+      long_desc <<~EOLD
+        Searches rubygems.org for InSpec plugins. Exits 0 on a search hit, 1 on user error,
+        2 on a search miss. PATTERN is a simple string; a wildcard will be added as
+        a suffix, unles -e is used.
+      EOLD
       option :all, desc: 'List all available versions, not just the latest one.', type: :boolean, aliases: [:a]
       option :exact, desc: 'Assume PATTERN is exact; do not add a wildcard to the end', type: :boolean, aliases: [:e]
+      # Wish we could hide this
+      option :'include-test-fixture', type: :boolean, desc: 'Internal use'
```
Comment (jquick) [inline lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb]: Try adding `hide: true` to your options here.

File: lib/inspec/plugin/v2/filter.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      instance.exclude?(plugin_name)
+    end
+
+    def exclude?(plugin_name)
+      # Currently, logic is very simple: is there an exact match?
+      # In the future, we might add regexes on names, or exclude version ranges
+      return false unless @filter_data[:exclude].detect { |e| e.plugin_name == plugin_name }
+
+      # OK, return entire data structure.
+      @filter_data[:exclude].detect { |e| e.plugin_name == plugin_name }
```
Comment (jquick) [inline lib/inspec/plugin/v2/filter.rb:24]: Not a huge fan of this as I would expect true/false. Due to the limited scope of this class though I think its OK.

**Review by jquick — COMMENTED** [review]
Nice addition @clintoncwolfe !

File: lib/inspec/plugin/v2/filter.rb (hunk 2 above)
Comment (clintoncwolfe) [reply lib/inspec/plugin/v2/filter.rb:24]: I agree.  Better ideas welcome!

File: lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb [hunk 3]
```diff
@@ -35,16 +36,31 @@ def list
       #                        inspec plugin search
       #==================================================================#
 
-      desc 'search [options] PATTERN', 'Searches rubygems.org for InSpec plugins. Exits 0 on a search hit, exits 2 on a search miss.'
+      desc 'search [options] PATTERN', 'Searches rubygems.org for plugins.'
+      long_desc <<~EOLD
+        Searches rubygems.org for InSpec plugins. Exits 0 on a search hit, 1 on user error,
+        2 on a search miss. PATTERN is a simple string; a wildcard will be added as
+        a suffix, unles -e is used.
```
Comment (jerryaldrichiii) [inline lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb]: `unles` should be `unless`

File: lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb [hunk 4]
```diff
@@ -342,8 +358,15 @@ def install_from_remote_gem_verson_preflight_check(plugin_name, requested_versio
...
 
-      def install_attempt_install(plugin_name)
+      # Rationale for RuboCop variance: This is a one-line method with heavy UX-focused error handling.
+      def install_attempt_install(plugin_name) # rubocop: disable Metrics/AbcSize
         installer.install(plugin_name, version: options[:version])
+      rescue Inspec::Plugin::V2::PluginExcludedError => ex
+        puts(red { 'Plugin on Exclusion List' } + " - #{plugin_name} is listed as an incompatible gem - refusing to install.")
+        puts "Rationale: #{ex.details.rationale}"
+        puts 'Exclusion list location: ' + File.join(Inspec.src_root, 'etc', 'plugin_filters.json')
+        puts 'If you disagree with this determination, please accept our apologies for the misunderstanding, and open an issue at https://github.com/inspec/inspec/issues/new'
```
Comment (jerryaldrichiii) [inline lib/plugins/inspec-plugin-manager-cli/lib/inspec-plugin-manager-cli/cli_command.rb:367]: Mmm. I like this.

File: lib/plugins/inspec-plugin-manager-cli/test/functional/inspec-plugin_test.rb [hunk 5]
```diff
@@ -513,6 +541,32 @@ def test_install_from_rubygems_latest_with_train_plugin
     refute_nil itf_line, 'train-test-fixture should now appear in the output of inspec list'
     assert_match(/\s*train-test-fixture\s+0.1.0\s+gem\s+/, itf_line, 'list output should show that it is a gem installation with version')
   end
+
+  def test_refuse_install_when_plugin_on_exclusion_list
+
+    # Here, 'inspec-core', 'inspec-multiserver', and 'train-tax-collector'
```
Comment (jerryaldrichiii) [inline lib/plugins/inspec-plugin-manager-cli/test/functional/inspec-plugin_test.rb]: Missed a `-` in `inspec-mutliserver`.

Also, below has extra spaces after `/ ` and on the last line before the period. Those are less critical though.

File: lib/plugins/inspec-plugin-manager-cli/test/unit/cli_args_test.rb [hunk 6]
```diff
@@ -26,7 +26,7 @@ def test_list_args
 
   def test_search_args
     arg_config = cli_class.all_commands['search'].options
-    assert_equal 2, arg_config.count, 'The search command should have 2 options'
+    assert_equal 3, arg_config.count, 'The search command should have 2 options'
```
Comment (jerryaldrichiii) [inline lib/plugins/inspec-plugin-manager-cli/test/unit/cli_args_test.rb]: I think we need to change the message to `3 options` as well.

File: test/unit/plugin/v2/installer_test.rb [hunk 7]
```diff
@@ -218,6 +218,24 @@ def test_install_a_plugin_from_a_path
     assert_includes entry.keys, 'installation_path', 'plugins.json should include installation_path key'
     assert_equal @plugin_fixture_src_path, entry['installation_path'], 'plugins.json should include correct value for installation path'
   end
+
+  def test_refuse_to_install_gem_whose_name_is_on_the_reject_list
+    ENV['INSPEC_CONFIG_DIR'] = File.join(@config_dir_path, 'empty')
+
+    # Here, 'inspec-core', 'inspec-multiserver', and 'train-tax-collector'
```
Comment (jerryaldrichiii) [inline test/unit/plugin/v2/installer_test.rb]: Same typos as above feedback. I think this is a copy/paste sort of propagation. I'll ignore further 'violations'.

**Review by jerryaldrichiii — CHANGES_REQUESTED** [review]
Looks great, only one blocking issue I think (options count). The rest is just typos and such.

File: lib/plugins/inspec-plugin-manager-cli/test/functional/inspec-plugin_test.rb (hunk 5 above)
Comment (clintoncwolfe) [reply lib/plugins/inspec-plugin-manager-cli/test/functional/inspec-plugin_test.rb]: Thanks.  I often add a space after a filename, to separate it from the period.  

I've changed it to quote the filename, that's clearer.

File: test/unit/plugin/v2/installer_test.rb [hunk 8]
```diff
@@ -218,6 +218,24 @@ def test_install_a_plugin_from_a_path
     assert_includes entry.keys, 'installation_path', 'plugins.json should include installation_path key'
     assert_equal @plugin_fixture_src_path, entry['installation_path'], 'plugins.json should include correct value for installation path'
   end
+
+  def test_refuse_to_install_gem_whose_name_is_on_the_reject_list
+    ENV['INSPEC_CONFIG_DIR'] = File.join(@config_dir_path, 'empty')
+
+    # Here, 'inspec-core', 'inspec-multi-server', and 'train-tax-collector'
+    # are the names of real rubygems.  They are not InSPec/Train plugins, though,
```
Comment (jerryaldrichiii) [inline test/unit/plugin/v2/installer_test.rb:226]: Whoops missed this one last time `InSPec` should be `InSpec`.

**Review by jerryaldrichiii — DISMISSED** [review]

**Review by jquick — COMMENTED** [review]
Thanks @clintoncwolfe

**Review by jquick — APPROVED** [review]

**Pull request merged**